main/src/library/ChannelImpl.flix

```diff
@@ -0,0 +1,374 @@
+//
+// Copyright 2018 Simon Meldahl Schmidt
+// Copyright 2018-2021 Jonathan Lindegaard Starup
+// Copyright 2021 Justin Fargnoli
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+namespace ChannelImpl {
+
+    ///
+    /// The tuple corresponds to (
+    ///   `id` - an unique id used to order the locking of sets of channels. This avoids deadlocks,
+    ///   `channelLock` - the lock of the channel,
+    ///   `buffered` - indicates whether a handoff is needed or not.
+    ///       If you try to put an element in a channel that's full, you wait until there's space,
+    ///   `maxQueueSize` - the maximum number of elements allowed in `deque`. This is `max(1, channel size)`,
+    ///   `deque` - a deque of the channel elements,
+    ///   `waitingGetters` - a set of conditions that are waiting for an element.
+    ///       This set is cleared after each new element,
+    ///   `waitingSetters` - a condition that can notify threads of available space in `deque`.
+    /// )
+    /// The reason for the difference between `waitingGetters` and `waitingSetters` is the select expression.
+    /// A thread can only wait for one condition which is a problem when a select expression requires waiting for a
+    /// set of channels. Therefore the thread makes a condition that represents the select expression and hands that to
+    /// every channel involved.
+    ///
```

Comment:
We should probably state that access to the mutable data structures, MutDeque and MutList are guarded by the channel lock, right?